python/ray/air/_internal/tensorflow_utils.py

```diff
@@ -112,12 +115,25 @@ def get_shape(dtype: Union[np.dtype, pa.DataType]) -> Tuple[int, ...]:
             shape += dtype.element_shape
         return shape
 
+    def get_tensor_spec(
+        dtype: Union[np.dtype, pa.DataType], *, name: str
+    ) -> tf.TypeSpec:
+        shape, dtype = get_shape(dtype), get_dtype(dtype)
+        # Batch dimension is always `None`. So, if there's more than one `None`-valued
+        # dimension, then the tensor is ragged.
+        is_ragged = sum(dim is None for dim in shape) > 1
+        if is_ragged:
+            type_spec = tf.RaggedTensorSpec(shape, dtype=dtype)
+        else:
+            type_spec = tf.TensorSpec(shape, dtype=dtype, name=name)
+        return type_spec
```

Comment:
Nice! Another option would be checking:
```python
if isinstance(dtype, TensorDtype) and dtype.is_variable_shaped() or isinstance(dtype, ArrowVariableShapedTensorType):
    # ragged
```
but what you have, using just the shape, is way cleaner IMO!